### PR TITLE
[WIP] Feature/ref without dbref

### PR DESCRIPTION
**Pull request by Nilz11**
See #1560

### CODE REVIEW

**Comment by Nilz11** [comment]
Any comment on this so far?

**Comment by alcaeus** [comment]
Sorry, I haven't gotten to it yet. I'll make some time for an initial review tonight.

**Comment by Nilz11** [comment]
Ok, just want to know if the general approach is ok like that. Still want to do testing and some minor modifications.

**Comment by alcaeus** [comment]
No specifics yet, just one thing I was wondering: do we really need to make a distinction between `ref` and `refWithDb`? For normal references, this isn't necessary since we'll have a `ClassMetadata` object tell us what DB the referenced object is in. The only time we'll need the `db` field in the reference is when there is no `targetDocument` set on the reference, in which case we can automatically add the `db` field as necessary. FWIW, the same would work for the `DbRef` strategy.

While I like the idea of storing references in documents that can be used in aggregation pipeline stages, I'm not entirely sure adding two new reference strategies is the way to go. Maybe we could reduce it to one, maybe we could just add `id`, `ref` and (optionally) `db` fields to the DbRef object for use in aggregation pipeline stages?

**Comment by Nilz11** [comment]
Ok I can change that appropriately, didn't know that there is a db field inside the class metadata.

So if no target document is set, we have the following fields: `id`, `ref`, `db` and the discriminator or `_doctrine_class_name`. Otherwise just id and ref.

**Comment by Nilz11** [comment]
Also I thought of providing a migration command using the native mongo driver, would be quite easy to implement and could migrate all relations inside database.

**Comment by Nilz11** [comment]
So depending on the release version and general release strategy, if it's 2.0 DbRef could be removed completely then with the need to migrate and in 1.x it could be deprecated.

**Comment by Nilz11** [comment]
One more comment about optional db field, in most cases refs are inside the same db, so if we just make one type, there would be always an unnecessary db value inside the refs without target document

**Comment by malarzm** [comment]
I know that multi-tenant apps aren't that popular but here goes my 2 cents:

> For normal references, this isn't necessary since we'll have a ClassMetadata object tell us what DB the referenced object is in. The only time we'll need the db field in the reference is when there is no targetDocument set on the reference

Please don't infer `db` from `ClassMetadata` blindly, I was working on a multi-tenant application where we weren't specifying db as it was dynamic (depending on context set during request handling) and were relying on default database set for `DocumentManager` (so in practice we had our own `DocumentManagerFactory` that was producing different instances for each tenant, but all document managers were sharing documents metadata, event listeners and so on, the only difference was that default database setting in configuration).

> One more comment about optional db field, in most cases refs are inside the same db, so if we just make one type, there would be always an unnecessary db value inside the refs without target document

I see "in most cases", just wanted to let you know that my app was outside of that "most cases" :) We were utilizing `ReferenceMany` with and without `targetDocument` targeting multiple other databases.

**Comment by Nilz11** [comment]
With the last comment I wanted to say, that it could make sense to have two types of refs one with db and one without. So if you have multiple databases you can use this type, otherwise just the normal one which leaves out the db field and so you do not have unnecessary information inside the database (even if you do not specify target document).

**Comment by Nilz11** [comment]
In the end that's not a decision I can make. My proposal would be to make everything similar to current implementation. Changing this behaviour later is easily possible

**Comment by malarzm** [comment]
I'm marking this for 1.2 so we could already deprecate DBRefs and remove them in 2.0

**Comment by alcaeus** [comment]
Preparing for a merge into 1.2, so let's solve a few issues.

> Please don't infer db from ClassMetadata blindly, I was working on a multi-tenant application where we weren't specifying db as it was dynamic (depending on context set during request handling) and were relying on default database set for DocumentManager (so in practice we had our own DocumentManagerFactory that was producing different instances for each tenant, but all document managers were sharing documents metadata, event listeners and so on, the only difference was that default database setting in configuration).

We have to trust our metadata. If the db property of two metadata instances is the same we can't but assume that they are stored in the same database. That being said, I'm fine with leaving that magic behavior out for now.

> I'm marking this for 1.2 so we could already deprecate DBRefs and remove them in 2.0

Given our last hangout, wen can still deprecate DBRef, but I wouldn't remove them in 2.0 just yet, since the migration could be quite painful.

@Nilz11 could you give this branch a rebase against master? Thanks!

File: lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php [hunk 1]
```diff
@@ -466,6 +468,30 @@ public function __construct($documentName)
...
 
     /**
+     * Helper method to get reference id of ref* type references
+     * @param array  $reference
+     * @param string $storeAs
+     * @return \MongoId
+     */
+    public static function getReferenceId($reference, $storeAs)
+    {
+        if ($storeAs === ClassMetadataInfo::REFERENCE_STORE_AS_REF_WITH_DB || $storeAs === ClassMetadataInfo::REFERENCE_STORE_AS_REF) {
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php]: This method should just use `getReferencePrefix`

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 2]
```diff
@@ -1161,7 +1166,7 @@ private function prepareQueryElement($fieldName, $value = null, $class = null, $
 
         // Prepare DBRef identifiers or the mapped field's property path
         $fieldName = ($objectPropertyIsId && ! empty($mapping['reference']) && $mapping['storeAs'] !== ClassMetadataInfo::REFERENCE_STORE_AS_ID)
-            ? $e[0] . '.$id'
+            ? $e[0] . '.'.ClassMetadataInfo::getReferencePrefix($mapping['storeAs']).'id'
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Spaces between strings and `.` please.

**Comment by alcaeus** [comment]
@Nilz11 gave your branch a quick rebase, fixed checkstyles and squashed commits. I'll take another look in the coming days and finish this up. Thanks for your help! 🍻

**Comment by alcaeus** [comment]
Superseded by #1623.

**Pull request closed without merging**